crosschain-resolver/contracts/L1Resolver.sol

```diff
@@ -4,20 +4,65 @@ pragma solidity ^0.8.17;
 import {EVMFetcher} from '@ensdomains/evm-verifier/contracts/EVMFetcher.sol';
 import {EVMFetchTarget} from '@ensdomains/evm-verifier/contracts/EVMFetchTarget.sol';
 import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+import "@ensdomains/ens-contracts/contracts/registry/ENS.sol";
+import {INameWrapper} from "@ensdomains/ens-contracts/contracts/wrapper/INameWrapper.sol";
 
 contract L1Resolver is EVMFetchTarget {
     using EVMFetcher for EVMFetcher.EVMFetchRequest;
     IEVMVerifier immutable verifier;
-    address immutable target;
+    ENS immutable ens;
+    INameWrapper immutable nameWrapper;
+    mapping(bytes32 => address) public targets;
     uint256 constant COIN_TYPE_ETH = 60;
-    uint256 constant RECORD_VERSIONS_SLOT = 1;
-    uint256 constant VERSINABLE_ADDRESSES_SLOT = 3;
-    uint256 constant VERSINABLE_HASHES_SLOT = 4;
-    uint256 constant VERSINABLE_TEXTS_SLOT = 11;
-
-    constructor(IEVMVerifier _verifier, address _target) {
-        verifier = _verifier;
-        target = _target;
+    uint256 constant RECORD_VERSIONS_SLOT = 0;
+    uint256 constant VERSINABLE_ADDRESSES_SLOT = 2;
+    uint256 constant VERSINABLE_HASHES_SLOT = 3;
+    uint256 constant VERSINABLE_TEXTS_SLOT = 10;
+
+    function isAuthorised(bytes32 node) internal view returns (bool) {
+        // TODO: Add support for
+        // trustedETHController
+        // trustedReverseRegistrar
+        // isApprovedForAll
+        // isApprovedFor
+        address owner = ens.owner(node);
+        if (owner != address(0) && owner == address(nameWrapper)) {
```

Comment:
You don't need to check that it's both nonzero and equal to a specific value - just the latter will suffice as long as you check the namewrapper address is nonzero in the constructor.

Reply:
I added the check so that I don't have to setup namewrapper in the test (namewrapper deployment kept throwing up an error). I can take it out when I add proper suport for other permissions

Reply:
It seems kind of important to actually test the namewrapper functionality too.

Reply:
@jefflau Do I have to set anything before instantiating new Wrapper? I created a valid ENS registry, registrar, and metadata service and passed to wrapper pfollowing this example](https://github.com/ensdomains/ens-contracts/blob/staging/test/wrapper/NameWrapper.js#L133)  but the deployment is reverting

```
    const ensFactory = await ethers.getContractFactory('ENSRegistry',signer);
    ens = await ensFactory.deploy();
    const ensAddress = await ens.getAddress()
    const baseRegistrarFactory = await ethers.getContractFactory('BaseRegistrarImplementation',signer);
    const baseRegistrar = await baseRegistrarFactory.deploy(ensAddress,ethers.namehash('eth'))
    const baseRegistrarAddress = await baseRegistrar.getAddress()
    await baseRegistrar.addController(signerAddress)
    const metaDataserviceFactory = await ethers.getContractFactory('StaticMetadataService',signer);
    const metaDataservice = await metaDataserviceFactory.deploy('https://ens.domains')
    const metaDataserviceAddress = await metaDataservice.getAddress()
    const wrapperFactory = await ethers.getContractFactory('NameWrapper',signer);
    wrapper = await wrapperFactory.deploy(
        ensAddress,
        baseRegistrarAddress,
        metaDataserviceAddress
    );
```

Any idea why it's happening?

```
  1) Crosschain Resolver
       "before all" hook for "should not allow non owner to set target":
     Error: missing revert data (action="estimateGas", data=null, reason=null, transaction={ "data": "0x60c06040818152346200030d57606082620067cc803803809162000024828562000312565b8339810103126200030d5781516001600160a01b038082169391908482036200030d57602092838101519082821682036200030d57850151928284168094036200030d57600080978154948760018060a01b0319963388821617855560248b5180958193863391167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e08a80a36302571be360e01b82527f91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e260048301525afa8015620003035788928491620002bd575b508951630f41a04d60e11b8152336004820152938492602492849291165af18015620002b3576200027f575b5091600693918697959360805260a05260055416176005557f93cdeb708b7545dc668eb9280176169d1c33cfd8ed6f04690a0bcc88a93fc4ae80845260018252600163fffeffff60a01b03198086862055848052858520558282526002858520620001966200018f82546200034c565b8262000389565b5583525220620001ab6200018f82546200034c565b600a626cae8d60e31b019055516164089081620003c48239608051818181610ad001528181610f0b015281816116da015281816117660152818161263c015281816127490152818161293a01528181612c1e015281816130900152818161490b01528181614e7901528181614f3f015281816150cd01528181615e7d01528181615f5a0152818161606701526162f6015260a051818181610c4c0152818161122d01528181611d9701528181611edb01528181612417015281816125b7015281816126c4015281816147e101526151de0152f35b8590813d8311620002ab575b62000297818362000312565b81010312620002a757386200011f565b8680fd5b503d6200028b565b87513d8a823e3d90fd5b8381939492503d8311620002fb575b620002d8818362000312565b81010312620002f75751908082168203620002f75787916024620000f3565b8280fd5b503d620002cc565b89513d85823e3d90fd5b600080fd5b601f909101601f19168101906001600160401b038211908210176200033657604052565b634e487b7160e01b600052604160045260246000fd5b90600182811c921680156200037e575b60208310146200036857565b634e487b7160e01b600052602260045260246000fd5b91607f16916200035c565b601f821162000396575050565b6000908152601f60208220920160051c8201915b828110620003b757505050565b818155600101620003aa56fe6080604052600436101561001257600080fd5b60
```

Reply:
I also added reverse registrar and set public resolver against it but still throwing the same error.

```
    const reverseRegistrarFactory = await ethers.getContractFactory('ReverseRegistrar',signer);
    const reverseRegistrar = await reverseRegistrarFactory.deploy(ensAddress)
    const reverseRegistrarAddress = await reverseRegistrar.getAddress()
    await ens.setSubnodeOwner(EMPTY_BYTES32, labelhash('reverse'), signerAddress)
    await ens.setSubnodeOwner(ethers.namehash('reverse'),labelhash('addr'), reverseRegistrarAddress)
    await ens.setSubnodeOwner(EMPTY_BYTES32, labelhash('eth'), signerAddress)
    await ens.setSubnodeOwner(ethers.namehash('eth'), labelhash('foo'), signerAddress)
    const publicResolverFactory = await ethers.getContractFactory('PublicResolver',signer);
    const publicResolver = await publicResolverFactory.deploy(
      ensAddress,
      '0x0000000000000000000000000000000000000000',
      '0x0000000000000000000000000000000000000000',
      reverseRegistrarAddress,
    )
    const publicResolverAddress = await publicResolver.getAddress()
    await reverseRegistrar.setDefaultResolver(publicResolverAddress)
```

Reply:
Have you tracked down which statement in the hook is causing it?

Reply:
I just removed all codes inside constructor but it's still throwing revert error on deploy. I am guessing it's happening at some other super classe's constructor.

This is the version I made it deployable (verifying that I am passing the right number of arguments with correct types). Once I start adding other superclass it either reverts or compilation error with missing implementation.

```
contract NameWrapper is
    Ownable,
    ERC20Recoverable,
    ReverseClaimer
{
    ENS public immutable ens;
    IBaseRegistrar public immutable registrar;
    IMetadataService public metadataService;
    constructor(
        ENS _ens,
        IBaseRegistrar _registrar,
        IMetadataService _metadataService
    ) ReverseClaimer(_ens, msg.sender) {
        ens = _ens;
        registrar = _registrar;
        metadataService = _metadataService;
    }
}
```

Reply:
This is the state where I can deploy NameWrapper without revert. It's very strange because `setUpgradeContract` isn't called by any smart contract internally.

<img width="1409" alt="Screenshot 2023-11-03 at 16 26 55" src="https://github.com/ensdomains/evmgateway/assets/2630/235a8d30-6414-47e4-a171-55572021f88c">

Reply:
Another issue is the fact that the test setting isn't catching any revert message so it's almost impossible to dianoze what the issue is.

```
       should allow wrapped owner to set target:
     Error: missing revert data (action="estimateGas", data=null, reason=null, transaction={ "data": "0xfca247ac4ca938ec1b323ca71c4fb47a712abb68cce1cabf39ea4d6789e42fbc1f95459b0000000000000000000000007b307aea9588b11ea026c83df61a1694933891220000000000000000000000000000000000000000000000000000000005f5e100", "from": "0x7b307aea9588b11Ea026C83df61A169493389122", "to": "0xbc3e9A0f522CF01A39B1445D5a7EE3AA2aE54596" }, invocation=null, revert=null, code=CALL_EXCEPTION, version=6.8.0)
```

Reply:
Have you narrowed it down to one line in the beforeEach function? Make sure it's actually the thing you think it is that's causing the error.

Reply:
I am quite certain that it's the deployer contract that's throwing an error

I put console.log like this
```
    console.log(4, {ensAddress,baseRegistrarAddress, metaDataserviceAddress})
    const wrapperFactory = await ethers.getContractFactory('NameWrapper',signer);
    console.log(5)
    await provider.send('evm_mine', []);
    console.log(6)
    wrapper = await wrapperFactory.deploy(
      ensAddress,
      baseRegistrarAddress,
      metaDataserviceAddress
    );
    console.log(7, {wrapperAddress})
```

 and it only prints up to 6

```
4 {
  ensAddress: '0x5c715383C0412139C0FBc7307d297c212C3eF2Dd',
  baseRegistrarAddress: '0x8841E0992927E920D6Deea6BB2D5a4faF969cf48',
  metaDataserviceAddress: '0x8779dccf8349136ACC1C4ec28c50524d8B5ca2ED'
}
5
6
    1) "before all" hook for "should not allow non owner to set target"


  0 passing (635ms)
  1 failing

  1) Crosschain Resolver
       "before all" hook for "should not allow non owner to set target":
     Error: missing revert data (action="estimateGas", data=null, reason=null, transaction={ "data": "0x60c0604081815234620002f857606082620065df8038038091620000248285620002fd565b833981010312620002f85781516001600160a01b03808216939190848203620002f85760209283810151908282168203620002f85785015192828416809403620002f857600080978154948760018060a01b0319963388821617855560248b5180958193863391167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e08a80a36302571be360e01b82527f91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e260048301525afa8015620002ee5788928491620002a8575b508951630f41a04d60e11b8152336004820152938492602492849291165af180156200029e576200026a575b5091600693918697959360805260a05260055416176005557f93cdeb708b7545dc668eb9280176169d1c33cfd8ed6f04690a0bcc88a93fc4ae80845260018252600163fffeffff60a01b03198086862055848052858520558282526002858520620001966200018f825462000337565b8262000374565b5583525220620001ab6200018f825462000337565b600a626cae8d60e31b019055516162309081620003af8239608051818181610ad001528181610f0b015281816116da015281816117660152818161275601528181612a4601528181612eb80152818161473301528181614ca101528181614d6701528181614ef501528181615ca5015281
```

Reply:
Okay; have you tried commenting out parts of the constructor in the copy of the contract it's using, to see which line causes the error?

Reply:
I commented out all the code in the constructor but was still failing. It's strange that none of the constructor code (including constructor of the super class) calls neither `approve` nore `setUpgradeContract` and my test code don't call these functions either.